chalk-ir/src/visit.rs

```diff
@@ -0,0 +1,421 @@
+//! Traits for visiting bits of IR.
+use std::fmt::Debug;
+
+use crate::{
+    BoundVar, DebruijnIndex, DomainGoal, Goal, InferenceVar, Interner, Lifetime, LifetimeData,
+    PlaceholderIndex, ProgramClause, Ty, TyData, WhereClause,
+};
+
+mod binder_impls;
+mod boring_impls;
+pub mod visitors;
+
+pub use visitors::VisitExt;
+
+pub trait VisitResult: Sized {
```

Comment:
Let's add a comment:

---

A "result type" that can be returned from a visitor. Visitors pick an appropriate result type depending on what sort of operation they are doing. A common choice is `FindAny`, which indicates that the visitor is searching for something and that the visitor should stop once it is found.